FIX properly report `n_iter_` in case of fallback from Newton-Cholesky to LBFGS

ogrisel opened the pull request:
Fix for a small bug discovered while reviewing #28840: whenever the Newton-Cholesky solver of LogisticRegression would fall back to LFGS, the reported `n_iter_` attribute would be left to zero.

I made it such that any completed iteration from the Newton-Cholesky solver would be subtracted from `max_iter` before calling LBFGS, and then report the sum of the numbers of iterations completed by the two solvers in the end. In practice, this does not seem to change anything because when Hessian conditioning problem always happens during the first iteration in my experiments with low regularized, rank deficient problems that typically trigger the LBFGS fallback mechanism.

Note that I find that the LBFGS fallback warning quite annoying whenever it is triggered while tuning the regularization level (e.g. using `LogisticRegressionCV` or `RandomizedSearchCV`). I have the feeling that this should be a regular verbose print instead, but we can tackle that in a separate PR.

ogrisel committed:
FIX properly report n_iter in case of Newton-Cholesky to LBFGS fallback

ogrisel committed:
DOC add a changelog entry

ogrisel committed:
Make the problem even more ill-conditioned

lorentzenchr commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -2374,3 +2374,43 @@ def test_multi_class_deprecated():
+    assert n_iter_lbgs >= 1
+
+    # Check that the Newton-Cholesky solver raises a warning and falls back to
+    # LBFGS. This should also converges in the same number of iterations as the
+    # previous call since the Newton-Cholesky solver should trigger the
+    # fallback before completing the first iteration.
```
```suggestion
    # LBFGS. This should converge with the same number of iterations as the
    # above call of lbfgs since the Newton-Cholesky triggers the fallback
    # before completing the first iteration, for the problem setting at hand.
```

ogrisel replied:
LGTM, I will accept the suggestion once the current CI run has completed to check if the way the test is written is portable enough for the different BLAS / scipy / OS kinds and versions we test against.

lorentzenchr reviewed (approved):
Thanks for spotting and fixing this bug.
I also would like to see improvements for the fallback case.

ogrisel committed:
Update sklearn/linear_model/tests/test_logistic.py

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

ogrisel committed:
[all random seeds]
test_newton_cholesky_fallback_to_lbfgs

ogrisel commented:
The `global_random_seed` parametrized test pass for all seeds locally.

Unfortunately this is not the case on our `macOS pylatest_conda_mkl_no_openmp` runner, even after having decreased the regularization and increased the data width.

There is a failure with `global_random_seed=6` on that host.

```python
        # Trying to fit the same model again with a small iteration budget should
        # therefore raise a ConvergenceWarning:
        lr_nc_limited = LogisticRegression(
            solver="newton-cholesky", C=C, max_iter=n_iter_lbgs - 1
        )
>       with pytest.warns(LinAlgWarning, match="ill-conditioned Hessian matrix"):
E       Failed: DID NOT WARN. No warnings of type (<class 'scipy.linalg._misc.LinAlgWarning'>,) were emitted.
E       The list of emitted warnings is: []
```

I pushed a new commit with the `[all random seeds]` commit message for that test to see there is only a problem with this CI environment or if the test is also brittle on other environment.

ogrisel commented:
Ok, so the warning is never raised for `macOS pylatest_conda_mkl_no_openmp` whatever the choice of the random seed, while it is always raised (as expected) for all the seeds in all the other testing environments. So the problem is not related to the statistical properties of the test problem.

ogrisel committed:
Fix typo

ogrisel committed:
Improve test:

- use ignore_warnings(category=LinAlgWarning) for now;
- store n_iter values in local variables to help CI-log debugging;
- fix final assertion.

ogrisel committed:
[azure parallel] [all random seeds] test_newton_cholesky_fallback_to_lbfgs

ogrisel commented:
I tried to look into the scipy version / BLAS version and so, but I could not find anything that would explain why the warning was not raised only in `macOS pylatest_conda_mkl_no_openmp` while it works for other environments (with or without MKL, with older and newer versions of scipy and so on).

ogrisel commented:
Ok, so it seems to be a bug in the propagation of `LinAlgWarning` but the iteration logic is still somehow respected...

I will push another commit to clean-up the commented out code and keep the `ignore_warnings` thing for now.

ogrisel committed:
Clean-up

glemaitre commented on doc/whats_new/upcoming_changes/sklearn.linear_model/30100.fix.rst:
```diff
@@ -0,0 +1,5 @@
+  accept `solver="newton-cholesky"` now report the correct number of iterations
+  when they fall back to the `"lbfgs"` solver because of a rank deficient
+  Hessian matrix.
+  By :user:`Olivier Grisel <ogrisel>`.
```
I need to finish my PR to solve this issue with the full stop :)

```suggestion
  By :user:`Olivier Grisel <ogrisel>`
```

glemaitre committed:
Update doc/whats_new/upcoming_changes/sklearn.linear_model/30100.fix.rst

adrinjalali commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -2374,3 +2374,47 @@ def test_multi_class_deprecated():
+            lr_nc_limited.fit(X, y)
+            n_iter_nc_limited = lr_nc_limited.n_iter_[0]
+
+    # XXX: Is this a one-off error? Shouln't it be equal to
+    # lr_nc_limited.max_iter instead?
+    assert n_iter_nc_limited == lr_nc_limited.max_iter - 1
```
does this need to be resolved?

ogrisel replied:
I am not sure if this is a bug or not. We would need to evaluate how other scikit-learn estimator behave in that respect (and maybe do a common test to enforce consistency).

But it's independent of the particular bug this PR fixes.

lorentzenchr replied:
I would remove the comment.
And tbh (what follows is a frank statement of my personal opinion), I don't consider this a bug. We trigger a convergence warning anyway. In this exceptional case, the solver tries not to abandon completely, but to give a meaningful result (the coefficients). I don't really care how to exactly count the numbers of iterations in such a scenario, it's screwed anyway. Don't we have more important things/inconsistencies to fix?

ogrisel replied:
Point taken.

ogrisel commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -2374,3 +2374,47 @@ def test_multi_class_deprecated():
+            lr_nc_limited.fit(X, y)
+            n_iter_nc_limited = lr_nc_limited.n_iter_[0]
+
+    # XXX: Is this a one-off error? Shouln't it be equal to
+    # lr_nc_limited.max_iter instead?
```
```suggestion
```

ogrisel committed:
Remove comment about max_iter - 1
